sapicore/utils/io.py

```diff
@@ -0,0 +1,270 @@
+                    attr_tensor = getattr(self.component, attr)
+                    expanded_dimensions = attr_tensor[None, :].shape
+                    self.chunk_size[attr] = 0
+
+                    hf.create_dataset(
+                        name=attr,
+                        shape=expanded_dimensions,
+                        compression="lzf",
+                        chunks=True,
+                        maxshape=tuple([s if j > 0 else None for j, s in enumerate(expanded_dimensions)]),
+                    )
+
+                # write configurable properties ONCE for future reference, e.g. during tensorboard loggable.
+                if hasattr(self.component, "_config_props"):
+                    for prop in getattr(self.component, "_config_props"):
+                        # create a dataset for this configurable property.
+                        hf.create_dataset(name=prop, shape=attr_tensor.shape, compression="lzf")
+
+                        # extract value to be logged and move to CPU if applicable.
+                        value = getattr(self.component, prop)
+                        if type(value) is Tensor:
+                            # if tensor, move to cpu.
+                            value = value.cpu()
+
+                        try:
+                            # store the configurable value or tensor in the dataset.
+                            hf[prop][:] = value
+                        except TypeError:
+                            # skip field if not compatible with HDF5 (e.g., a method reference).
+                            pass
+
+    def save_outputs_hook(self) -> Callable:
+        def fn(_, __, output):
+            """Updates the output cache and appends to destination HDF5 file if cache is full or simulation ends."""
+            # increment iteration counter.
+            self.iterations += 1
+
+            # add current outputs to this data accumulator instance's cache dictionary, whose values are tensors.
+            for attr in self.attributes:
+                if attr not in self.cache.keys():
+                    # the cache dictionary is empty because this is the first iteration post-dump or in general.
+                    self.cache[attr] = output[attr][None, :]  # expand output by one dimension (zero axis) to fit.
+                else:
+                    # vertically stack output attribute to cache tensor at the appropriate key.
+                    self.cache[attr] = torch.vstack([self.cache[attr], output[attr][None, :]])
+
+            # calculate cache size (MB).
+            cache_size = sum([v.element_size() * v.nelement() for _, v in self.cache.items()]) / 1000000.0
+
+            # if memory usage exceeds the limit given or reached the end of the run, append cached content to HDF5.
+            if cache_size > CHUNK_SIZE or self.iterations == self.entries:
+                with hdf.File(os.path.join(self.log_dir, self.component.identifier + ".h5"), "a") as hf:
```

Comment:
We should find a way to not have to open and close the file for every forward call because that would be slow af. E.g. convert the class into a contextmanager itself. Or give it a open/close function users can call. But that can wait for later. But, this is typically how I'd handle instead of having a `entries` prop to check against, which users generally won't actually know what may be, even if they have a guess. We keep the file open and then flush at the end when the contextmanager exits and then close the file so no data is left unwritten.

We should also add a register and release method so users can manually register and unregister the forward callback.

Reply:
Fantastic suggestion. `entries` irked me from the get go but I didn't see a clear way around it. Saving for later, but definitely open an issue for this after we merge.